File: 8_Sorting2/1MergeSort/MergeSort.py
def merge(arr , l, mid, h):
    temp=[]
    left=l      # agr left use na krke l hi use karenge to l ki value change hogi aage aane wale 
    r =mid+1    # operations me aur fir temp se arr me add krte time l ki value galat hogi
    while(left<=mid and r<=h):
        if(arr[left]<=arr[r]):
            temp.append(arr[left])
            left+=1

        else:
            temp.append(arr[r])
            r+=1

    while(left<=mid):
        temp.append(arr[left])
        left+=1

    while(r<=h):
        temp.append(arr[r])
        r+=1

    # now adding elements of temp to actual arr
    for i in range (l,h+1):
        arr[i]= temp[i-l]



def mergeSort(arr,l,h):
    mid = (l+h)//2

    if(l>=h):
        return
    
    mergeSort(arr,l,mid)
    mergeSort(arr,mid+1,h)
    merge(arr,l,mid,h)






arr  = [3,1,2,4,1,5,6,2,4]

mergeSort(arr,0,len(arr)-1)
print(arr)

File: REVISION/Sorting2/MergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;


void merge( int arr[], int l , int mid , int r){
 
    vector<int> temp;

    int n1 = l;
    int n2 = mid +1;
    while( n1 <= mid && n2 <=r){
        if( arr[n1]<= arr[n2]){
            temp.push_back(arr[n1]);
            n1++;
        }
        else{
            temp.push_back(arr[n2]);
            n2++;
        }
    }

    if( n1 > mid){
        while(n2 <=r){
            temp.push_back(arr[n2]);
            n2++;
        }
    }

    if( n2 > r){
        while(n1 <= mid){
            temp.push_back(arr[n1]);
            n1++;
        }
    }
    // copy back to original array //

    for(int i=l; i<=r; i++){
        arr[i] = temp[i-l];
    }

}



void mergeSort( int arr[],int l, int r){

    if(l>=r){
        return;
    }

    int mid = l+ (r-l)/2 ;

    mergeSort(arr, l, mid);
    mergeSort(arr, mid+1, r);

    merge(arr, l, mid , r);

    
}


int main(){

    int arr[] = {5,4,3,2,3,1};
    int n = sizeof(arr)/ sizeof(arr[0]);

    mergeSort(arr, 0, n-1);

    for( int i=0; i<n; i++){
        cout << arr[i] << " ";
    }

}